### ISSUE TITLE
Profit tracking...

### ISSUE THREAD

**Issue by Technorocker** [sammchardy/python-binance#250]
Hey everyone. Does anyone have a bit of code they would feel like sharing for keeping track of each order placed profit wise? I've been writing a bot for a while now and thought I had it working off of the buy or sell order price when it makes an order(minus trading fees) but last night when I thought I had it working my bot said I  made 230 worth of btc profit over night but my actual account says im 10 dollars down. So I'm not keeping track of it properly. If anyone could share how they track each orders profit or loss that would be very appreciated.

**Comment by gdm-willgu** [comment]
It'd be more meaningful to track your `portfolio` P&L. Your executed order changes your position, but for any given time, your know that you have `X` amount of BTC, `Y` amount of ETH, etc.

All you need is the market price of each crypto at any given point, so that you have mark-to-market values of all your positions. Adding them up would give you the value of your `portfolio`.

Any changes to this value over time would be your actual P&L. If  you don't care about real-time P&L, then as long as your have order history, you can re-calculate your P&L history. It should be fairly simple given that you understand how it works.

**Comment by Technorocker** [comment]
Thanks for the reply. Up until now I was basically taking the order price i put in and taking the difference of the last order price and calculating the profit as if I had 1 full btc. so it would tell me what the bot would be making with direct market price. Now I've adjusted it to look at my balance right after each successive order and breaking it down to what my amount ordered would be with a little bit of code to take off the 0.1% fee. 

This is my first time coding so its a learning process. Im sure experienced coders would shake their head if they seen my monstrosity of code structure and formatting lol

**Comment by Roibal** [comment]
Can you post up your code, Technorocker?

**Comment by Technorocker** [comment]
Well... I'll share to profit tracking bit of code i have. I havent really seen anyone on here post their whole code yet. but here is the part of my sell order where i try to track my profit from each order. The buy side of my ocde is pretty much the same code jsut when it says lastsellprice its lastbuyprice

```
# Check order to be filled
if mode == 'live':
	order = client.order_limit_sell(symbol=pair, quantity=maxqty, price=orderbook['bidPrice'])
	while True:
		checkorder = Orders.checkorderstatus(pair, order['orderId'])
		if checkorder == "FILLED":
			print('Sell order has gone through')
			if startorder is True or lastbuyprice is None:
				print('Startorder was true or lastprice was none. Setting lastsellprice')
				pass
			elif lasttrend != 'uptrend':
				pricediff = Balance.profit_calculator(lastbuyprice, sellprice)
				print('price difference is:', abs(pricediff))
				print('Maxqty was', maxqty)
				btc_after_fees = Balance.btc_value_after_fees(maxqty)
				realprofit = abs(pricediff) * btc_after_fees
				print('Realprofit is:', realprofit)
				checkbal = float(Balance.assetbalance(asset2))
				if sellprice < lastbuyprice:
					cash = checkbal - abs(realprofit)
					print(Fore.RED + 'last buy price was', lastbuyprice)
					print(Fore.RED + 'bought for', sellprice, 'and lost', Balance.format(realprofit),
	                   				      'profit')
					print(Fore.RED + 'Cash', Balance.format(cash))
					playsound('fail.mp3')
				else:
					cash = checkbal + abs(realprofit)
					print(Fore.GREEN + 'last buy price was', lastbuyprice)
					print(Fore.GREEN + 'sold for', sellprice, 'and made ', Balance.format(realprofit),
					      'profit')
					print(Fore.GREEN + 'Cash is now:', cash)
					playsound('Cheering.mp3')
				print('Total cash is now', cash)

			lastsellprice = sellprice
			startorder = False
			break
		else:
			#print('Not filled yet')
			if float(Binancewebsocket.btcres_array[0]['p']) < float(order['price']) - ((float(order['price']) * 0.2) / 100):
				print('Price went too far down')
				Orders.cancelorder(pair, order['orderId'])
				break
			time.sleep(2)
```

I use a websocket for btc/usdt prices.

**Comment by Roibal** [comment]
I'm considering uploading a bit of my codes as well

**Comment by Roibal** [comment]
@Technorocker  : 

here is a github fork of /examples/ including a trading bot I have developed, next task is to incorporate profit tracking - Maybe we can collaborate. Current focus is adding functionality to current wrapper, such as accessing/collecting/displaying data. 

I will eventually submit a pull-request to main python-binance master. 

[Roibal Fork of Python-Binance (RoibalBot.py is main Bot) ](https://github.com/Roibal/python-binance/tree/master/examples)

[Youtube video explaining functionality of RoibalBot.py example cryptocurrency algo trading bot](https://youtu.be/Y4RwD1OGz4c)

**Comment by Technorocker** [comment]
@Roibal I am watching your youtube video and I doubt I can be of any assistance with a blockchain engineer. As I mentioned this is my first attempt at coding. I took a python course on coursera just to figure out how to communicate with API's. I didnt know about the self. attribute in python like This in java till i was halfway through building my code. My lil bot is setup to handle various coins but I've really been focusing on BTC/USDT specifically. figuring that once I got that working properly I could incorporate running multiple bots on all 5 different pairs of coins that binance has. I will setup a github repo with my code in the next week here and I will share it. But be warned... as an engineer youre going to just shake you head. lol I am definitely up for swapping ideas and code so we all can make a few bucks on a working model. I've been scratching my head about this for 3 months since I started this lil hobby project. I thought I had it working at some point and made some changes to make it more friendly to some close friends to pass out to but messed up the code and went back to the drawing board. now I've probably tried over a dozen very basic ways to try and catch patterns in the candlesticks.
